ui/src/saros/ui/browser/SWTBrowser.java

```diff
@@ -0,0 +1,199 @@
+        () -> {
+          return browser.getUrl();
+        });
+  }
+
+  @Override
+  public boolean loadUrl(final String url, final int timeout) {
+    browser.getShell().open();
+
+    this.state = ReadyState.LOADING;
+    browser.addProgressListener(readyStateListener);
+
+    this.syncExec(
+        () -> {
+          browser.setUrl(url);
+        });
+    boolean successful = waitForPageLoaded(timeout);
+
+    browser.removeProgressListener(readyStateListener);
+
+    return successful;
+  }
+
+  private boolean waitForPageLoaded(final int timeout) {
+    AtomicBoolean waiting = new AtomicBoolean(true);
+    Display display = browser.getDisplay();
+
+    // Wake up the ui thread after the timeout exceeded
+    Runnable wakeUpTimeout =
+        () -> {
+          // avoid execution if url is already loaded
+          if (state != ReadyState.COMPLETE) {
+            waiting.set(false);
+            display.wake();
+          }
+        };
+
+    display.timerExec(timeout, wakeUpTimeout);
+
+    this.syncExec(
+        () -> {
+          while (waiting.get() && state != ReadyState.COMPLETE) {
+            if (!display.readAndDispatch()) display.sleep();
+          }
+        });
+
+    return this.state == ReadyState.COMPLETE;
+  }
+
+  @Override
+  public void loadHtml(final String html) {
+    this.syncExec(() -> browser.setText(html));
+  }
+
+  @Override
+  public Object evaluate(final String jsCode) {
+    return this.syncExec(
+        () -> {
+          Object rv = browser.evaluate(jsCode);
+          System.out.println(rv);
+          return rv;
+        });
+  }
+
+  @Override
+  public boolean execute(String jsCode) {
+    return this.syncExec(
+        () -> {
+          return browser.execute(jsCode);
+        });
+  }
+
+  @Override
+  public void addBrowserFunction(final AbstractJavascriptFunction function) {
+    this.syncExec(
+        () -> {
+          BrowserFunction swtFunction =
+              new BrowserFunction(browser, function.getName()) {
+                @Override
+                public Object function(Object[] arguments) {
+                  return function.function(arguments);
+                }
+              };
+          return swtFunction;
+        });
+  }
+
+  @Override
+  public void runOnDisposal(Runnable runnable) {
+    runOnDisposalList.add(runnable);
+  }
+
+  private <V> V syncExec(final NoCheckedExceptionCallable<V> callable) {
+    final AtomicReference<V> r = new AtomicReference<V>();
+    final AtomicReference<RuntimeException> exception = new AtomicReference<RuntimeException>();
+    browser
+        .getDisplay()
+        .syncExec(
+            () -> {
+              try {
+                r.set(callable.call());
+              } catch (RuntimeException e) {
+                exception.set(e);
+              }
+            });
+
+    if (exception.get() != null) {
+      throw exception.get();
+    }
+    return r.get();
+  }
+
+  private void syncExec(final Runnable runnable) {
```

Comment:
syncExec( () -> {runnable.run(); return null;});

Reply:
What do mean?

Reply:
This one-liner is a replacement for the current method body.

Reply:
Ah, you mean `SWTUtils.runSWTSync(() -> {runnable.run(); return null;})` ? Yes, should be the more convenient solution.

Reply:
You do not have access to SWTUtils. Simply insert the code, it will call the syncExec method of this class.

Reply:
Sorry, it is to hot. I forgot that I already defined another syncExec version.
**Done**